packages/evm/src/eof/container.ts

```diff
@@ -0,0 +1,367 @@
+import {
+  CODE_MIN,
+  CODE_SIZE_MIN,
+  CONTAINER_MAX,
+  CONTAINER_MIN,
+  CONTAINER_SIZE_MIN,
+  FORMAT,
+  INPUTS_MAX,
+  KIND_CODE,
+  KIND_CONTAINER,
+  KIND_DATA,
+  KIND_TYPE,
+  MAGIC,
+  MAX_HEADER_SIZE,
+  MAX_STACK_HEIGHT,
+  OUTPUTS_MAX,
+  TERMINATOR,
+  TYPE_DIVISOR,
+  TYPE_MAX,
+  TYPE_MIN,
+  VERSION,
+} from './constants.js'
+import { EOFError, validationError } from './errors.js'
+import { ContainerSectionType, verifyCode } from './verify.js'
+
+import type { EVM } from '../evm.js'
+
+export enum EOFContainerMode {
+  Default, // Default container validation
+  Initmode, // Initmode container validation (for subcontainers pointed to by EOFCreate)
+  TxInitmode, // Tx initmode container validation (for txs deploying EOF contracts)
+}
+
+class StreamReader {
+  private data: Uint8Array
+  private ptr: number
+  constructor(stream: Uint8Array) {
+    this.data = stream
+    this.ptr = 0
+  }
+
+  readBytes(amount: number, errorStr?: string) {
+    const end = this.ptr + amount
+    if (end > this.data.length) {
+      validationError(EOFError.OutOfBounds, this.ptr, errorStr)
+    }
+    const ptr = this.ptr
+    this.ptr += amount
+    return this.data.slice(ptr, end)
+  }
+
+  readUint(errorStr?: string) {
+    if (this.ptr >= this.data.length) {
+      validationError(EOFError.OutOfBounds, this.ptr, errorStr)
+    }
+    return this.data[this.ptr++]
+  }
+
+  verifyUint(expect: number, errorStr?: string) {
+    if (this.readUint() !== expect) {
+      validationError(EOFError.VerifyUint, this.ptr - 1, errorStr)
+    }
+  }
+
+  readUint16(errorStr?: string) {
+    const end = this.ptr + 2
+    if (end > this.data.length) {
+      validationError(EOFError.OutOfBounds, this.ptr, errorStr)
+    }
+    const ptr = this.ptr
+    this.ptr += 2
+    return new DataView(this.data.buffer).getUint16(ptr)
+  }
+
+  getPtr() {
+    return this.ptr
+  }
+
+  // Get the remainder bytes of the current stream
+  readRemainder() {
+    return this.data.slice(this.ptr)
+  }
+
+  isAtEnd() {
+    return this.ptr === this.data.length
+  }
+}
+
+// TODO add initcode flags (isEOFContract)
+// TODO validation: mark sections as either initcode or runtime code to validate
+
+class EOFHeader {
+  typeSize: number
+  codeSizes: number[]
+  containerSizes: number[]
+  dataSize: number
+  dataSizePtr: number // Used to edit the dataSize in RETURNCONTRACT
+  buffer: Uint8Array
+
+  private codeStartPos: number[]
+
+  constructor(input: Uint8Array) {
```

Comment:
I guess adding a simple debug logger `evm:eof` with a few strategic occasions for debug output like "Creating EOF header with 4 code sections and ..." would help a lot for making things more explicit and debug and analyze on various (test) situations and scenarios.

Doesn't need to be extensive on first round (so, e.g. 4-6 debug messages already a good start).